### PR TITLE
Thread Affinity

### PR DESCRIPTION
**Pull request by tianjunwork**
Add new api -lp and -ss to manage encoder threads affinity.

Signed-off-by: Jun Tian <jun.tian@intel.com>

### CODE REVIEW

**Comment by tianjunwork** [comment]
Tested on both Windows and Ubuntu.

File: Source/Lib/Codec/EbEncHandle.c [hunk 1]
```diff
@@ -246,60 +251,157 @@ EbAsm GetCpuAsmType()
...
+                groupAffinity.Group = config_ptr->target_socket;
+                }
+            }
+        }
+#else
+    const char* PROCESSORID = "processor";
+    const char* PHYSICALID = "physical id";
+    int processor_id_len = strnlen_ss(PROCESSORID, 128);
+    int physical_id_len = strnlen_ss(PHYSICALID, 128);
+    if (processor_id_len < 0 || processor_id_len >= 128) return EB_ErrorInsufficientResources;
```
Comment (jbeich) [inline Source/Lib/Codec/EbEncHandle.c]: Clang 8.0 reports:
```c
Source/Lib/Codec/EbEncHandle.c:323:58: error: void function 'EbSetThreadManagementParameters' should not return a value [-Wreturn-type]
    if (processor_id_len < 0 || processor_id_len >= 128) return EB_ErrorInsufficientResources;
                                                         ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: Source/Lib/Codec/EbEncHandle.c [hunk 2]
```diff
@@ -246,60 +251,157 @@ EbAsm GetCpuAsmType()
...
+    if (physical_id_len < 0 || physical_id_len >= 128) return EB_ErrorInsufficientResources;
```
Comment (jbeich) [inline Source/Lib/Codec/EbEncHandle.c]: Clang 8.0 reports:
```c
Source/Lib/Codec/EbEncHandle.c:324:56: error: void function 'EbSetThreadManagementParameters' should not return a value [-Wreturn-type]
    if (physical_id_len < 0 || physical_id_len >= 128) return EB_ErrorInsufficientResources;
                                                       ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: Source/Lib/Codec/EbEncHandle.c [hunk 3]
```diff
@@ -246,60 +251,157 @@ EbAsm GetCpuAsmType()
...
+                        while (*p < '0' || *p > '9') p++;
+                        processor_id = strtol(p, NULL, 0);
+                    }
+                    if (strncmp(line, PHYSICALID, physical_id_len) == 0) {
+                        char* p = line + physical_id_len;
+                        while (*p < '0' || *p > '9') p++;
+                        socket_id = strtol(p, NULL, 0);
+                        if (socket_id < 0 || socket_id > 15) {
+                            close(fd);
+                            return EB_ErrorInsufficientResources;
```
Comment (jbeich) [inline Source/Lib/Codec/EbEncHandle.c]: Clang 8.0 reports:
```c
Source/Lib/Codec/EbEncHandle.c:355:29: error: void function 'EbSetThreadManagementParameters' should not return a value [-Wreturn-type]
                            return EB_ErrorInsufficientResources;
                            ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by EwoutH** [comment]
Options look great, thanks! 

Closes #92.

**Comment by EwoutH** [comment]
It looks like `-lp` isn't integrated in the CLI yet, `-lp` doesn't show up in the `-help`, and when trying to use it, it states `Unprocessed tokens: -lp`.

**Comment by EwoutH** [comment]
Did a scaling test with `-lp 12` to `-lp 1`. Raw results:
 
| Speed (fps) | enc-mode 3 | enc-mode 2 | enc-mode 1 |
|------------:|-----------:|-----------:|-----------:|
|          12 |      2,476 |      0,737 |      0,300 |
|          11 |      2,426 |      0,738 |      0,306 |
|          10 |      2,119 |      0,658 |      0,269 |
|           9 |      2,074 |      0,664 |      0,273 |
|           8 |      1,719 |      0,566 |      0,232 |
|           7 |      1,690 |      0,558 |      0,230 |
|           6 |      1,303 |      0,441 |      0,183 |
|           5 |      1,270 |      0,431 |            |
|           4 |      0,936 |      0,324 |            |
|           3 |      0,871 |      0,289 |            |
|           2 |      0,474 |      0,163 |            |
|           1 |       fail |       fail |       fail |

`-lp 1` failed on all enc-modes with a memory allocation error, which is strange because less threads should mean less memory usage. 
`Could not allocate enough memory for channel 1`

The benefits from adding a thread are much smaller than those from adding a core, as observable in following table (results are normalized to `-lp 2)`.

| Speed (%) | enc-mode 3 | enc-mode 2 |
|----------:|-----------:|-----------:|
|        12 |      87,1% |      75,4% |
|        11 |      93,1% |      82,3% |
|        10 |      89,4% |      80,7% |
|         9 |      97,2% |      90,5% |
|         8 |      90,7% |      86,8% |
|         7 |     101,9% |      97,8% |
|         6 |      91,6% |      90,2% |
|         5 |     107,2% |     105,8% |
|         4 |      98,7% |      99,4% |
|         3 |     122,5% |     118,2% |
|         2 |     100,0% |     100,0% |

The uneven thread counts add a core with one thread, while the even thread counts only add a second thread to an already utilized core.

If we sort from low to high and only look at the even thread counts (with fully utilized cores), we see pretty impressive core scaling. Note that the core scaling is better on higher enc-modes.

| Speed (%) | enc-mode 3 | enc-mode 2 |
|----------:|-----------:|-----------:|
|         2 |     100,0% |     100,0% |
|         4 |      98,7% |      99,4% |
|         6 |      91,6% |      90,2% |
|         8 |      90,7% |      86,8% |
|        10 |      89,4% |      80,7% |
|        12 |      87,1% |      75,4% |



- **Build:** e74b841 (2019-02-23)
- **Video:** Morocco (1080p 8-bit 4:2:0)
- **Settings:** `-n 120`
- **System:** Ryzen 5 1600 (6c/12t, AVX2), 16GB DDR4, Windows 10 64-bit
- **Command:** `FOR %a in (3,2,1) DO (FOR %b in (0,12,11,10,9,8,7,6,5,4,3,2,1) DO SvtAv1EncApp -i Morocco_1080p_YT_240,y4m -enc-mode %a -lp %b -n 120)`

[Full details](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=555390849).

**Comment by tianjunwork** [comment]
@jbeich and @EwoutH, Thanks. I will fix the Clang report and -lp 1 error.

File: Source/Lib/Codec/EbEncHandle.c [hunk 4]
```diff
@@ -246,60 +251,157 @@ EbAsm GetCpuAsmType()
...
+        }
+#else
+    const char* PROCESSORID = "processor";
+    const char* PHYSICALID = "physical id";
+    int processor_id_len = strnlen_ss(PROCESSORID, 128);
+    int physical_id_len = strnlen_ss(PHYSICALID, 128);
+    if (processor_id_len < 0 || processor_id_len >= 128) return EB_ErrorInsufficientResources;
+    if (physical_id_len < 0 || physical_id_len >= 128) return EB_ErrorInsufficientResources;
+    CPU_ZERO(&groupAffinity);
+    numGroups = 1;
```
Comment (jbeich) [inline Source/Lib/Codec/EbEncHandle.c]: Can you drop this line per https://github.com/intel/SVT-HEVC/pull/115? `numGroups` is already initialized to 1 on line#165 and later adjusted on line#358.

Comment (tianjunwork) [reply Source/Lib/Codec/EbEncHandle.c]: Sure, I will include your intel/SVT-HEVC#115 patch.

**Comment by EwoutH** [comment]
`-lp 1` now works fine for me! (Windows 10)

**Comment by tianjunwork** [comment]
> `-lp 1` now works fine for me! (Windows 10)

Thanks for confirm:)

**Comment by EwoutH** [comment]
@tianjunwork @jbeich Can we merge?

**Comment by tianjunwork** [comment]
> @tianjunwork @jbeich Ready to merge?

@hassount please review the code and merge.

File: Source/Lib/Codec/EbEncHandle.c [hunk 5]
```diff
@@ -340,7 +442,12 @@ void LoadDefaultBufferConfigurationSettings(
     uint32_t meSegW     = (((sequence_control_set_ptr->max_input_luma_width + 32) / BLOCK_SIZE_64) < 10) ? 1 : 10;
     uint32_t inputPic   = SetParentPcs(&sequence_control_set_ptr->static_config);
 
-    unsigned int coreCount = GetNumCores();
+    unsigned int coreCount = GetNumProcessors();
```
Comment (hassount) [inline Source/Lib/Codec/EbEncHandle.c]: The idea is for the coreCount variable to get the number of cored that is now assigned for the library. if -lp is 40, this should be set to 40 and not the number of cores that is on the machine. would you be able to link these ?
Also if there are two sockets on the system and the library is set to default, we will need the number of cores to reflect the amount of cores in a single socket and not on the full system.

File: Build/windows/generate_vs17.bat [hunk 6]
```diff
@@ -1,5 +1,5 @@
 :: Copyright(c) 2019 Intel Corporation 
 :: SPDX-License-Identifier: BSD-2-Clause-Patent
 
-cmake ../.. -G "Visual Studio 15 2017 Win64" -DCMAKE_ASM_NASM_COMPILER="yasm.exe"
+cmake ../.. "-GVisual Studio 15 2017 Win64" -DCMAKE_ASM_NASM_COMPILER="yasm.exe" -DCMAKE_CONFIGURATION_TYPES="Debug;Release"
```
[the remark this answers is not in the document]
Comment (tianjunwork) [reply Build/windows/generate_vs17.bat]: sure. will update.

File: Source/Lib/Codec/EbEncHandle.c (hunk 5 above)
Comment (tianjunwork) [reply Source/Lib/Codec/EbEncHandle.c]: Yes, below logic is to set coreCount to -lp and -ss config. See line 446-450.
Why with default setting, coreCount is amount of cores in a single socket and not on the full system? with default setting, coreCount should be 118 on 8180, right?

Comment (hassount) [reply Source/Lib/Codec/EbEncHandle.c]: thank you for the clarification, I agree with you for linux, but on windows it won't run on both sockets by default, so we would be allocating more memory than  needed. can we make it different from windows / linux ?

**Comment by hassount** [comment]
> Did a scaling test with `-lp 12` to `-lp 1`. Raw results:
> 
> Speed (fps)	enc-mode 3	enc-mode 2	enc-mode 1
> 12	2,476	0,737	0,300
> 11	2,426	0,738	0,306
> 10	2,119	0,658	0,269
> 9	2,074	0,664	0,273
> 8	1,719	0,566	0,232
> 7	1,690	0,558	0,230
> 6	1,303	0,441	0,183
> 5	1,270	0,431	
> 4	0,936	0,324	
> 3	0,871	0,289	
> 2	0,474	0,163	
> 1	fail	fail	fail
> `-lp 1` failed on all enc-modes with a memory allocation error, which is strange because less threads should mean less memory usage.
> `Could not allocate enough memory for channel 1`
> 
> The benefits from adding a thread are much smaller than those from adding a core, as observable in following table (results are normalized to `-lp 2)`.
> 
> Speed (%)	enc-mode 3	enc-mode 2
> 12	87,1%	75,4%
> 11	93,1%	82,3%
> 10	89,4%	80,7%
> 9	97,2%	90,5%
> 8	90,7%	86,8%
> 7	101,9%	97,8%
> 6	91,6%	90,2%
> 5	107,2%	105,8%
> 4	98,7%	99,4%
> 3	122,5%	118,2%
> 2	100,0%	100,0%
> The uneven thread counts add a core with one thread, while the even thread counts only add a second thread to an already utilized core.
> 
> If we sort from low to high and only look at the even thread counts (with fully utilized cores), we see pretty impressive core scaling. Note that the core scaling is better on higher enc-modes.
> 
> Speed (%)	enc-mode 3	enc-mode 2
> 2	100,0%	100,0%
> 4	98,7%	99,4%
> 6	91,6%	90,2%
> 8	90,7%	86,8%
> 10	89,4%	80,7%
> 12	87,1%	75,4%
> 
> 
> * **Build:** [e74b841](https://github.com/OpenVisualCloud/SVT-AV1/commit/e74b841465ff781781f93261f678ddacb8ba1720) (2019-02-23)
> * **Video:** Morocco (1080p 8-bit 4:2:0)
> * **Settings:** `-n 120`
> * **System:** Ryzen 5 1600 (6c/12t, AVX2), 16GB DDR4, Windows 10 64-bit
> * **Command:** `FOR %a in (3,2,1) DO (FOR %b in (0,12,11,10,9,8,7,6,5,4,3,2,1) DO SvtAv1EncApp -i Morocco_1080p_YT_240,y4m -enc-mode %a -lp %b -n 120)`
> 
> [Full details](https://docs.google.com/spreadsheets/d/1p3PJQMkyhIrXEL6MBBQiwTnWmGyHY_zuw7_XAD0def4/edit#gid=555390849).

**Comment by EwoutH** [comment]
@hassount I have some more detailed numbers scaling to 50 threads in https://github.com/OpenVisualCloud/SVT-AV1/issues/98#issuecomment-467661974

File: Config/Sample.cfg [hunk 7]
```diff
@@ -79,8 +79,8 @@ HmeLevel2SearchAreaInHeight     : 1  1          # Array containing search area h
 
 ====================== Platform Specific Flags ===============================
 AsmType                         : 1             # Assembly instruction set (0: Lowest optimization available, 1: Highest optimization available)
-UseRoundRobinThreadAssignment   : 0             # For Dual socket systems running windows OS on systems with > 32 physical processors. When enbled, allows the encoder to run on both sockets  (0= OFF, 1=ON )
-TargetSocket                    : 1             # For Windows based dual socket systems, this can specify which socket the encoder should start / run on (depending on whether UseRoundRobinThreadAssignment is set to 1 or 0) 0= Socket 0, 1=Socket 1 )
+LogicalProcessors               : 0             # The number of logical processor which encoder threads run on [0-N] (N is maximum number of logical processor)
```
Comment (kylophone) [inline Config/Sample.cfg]: What does 0 indicate?

Comment (tianjunwork) [reply Config/Sample.cfg]: It is explained in user guide "0, total number of logical processor". The program run on all the logical processors. More detailed usage how to use -lp and -ss can be found in user guide appendix A.2.

**Comment by kylophone** [comment]
Does `-lp` limit the number of threads, or just the number of physical cores used?

File: Source/Lib/Codec/EbEncHandle.c (hunk 5 above)
Comment (tianjunwork) [reply Source/Lib/Codec/EbEncHandle.c]: Fixed on windows, please review.

**Comment by tianjunwork** [comment]
> Does `-lp` limit the number of threads, or just the number of physical cores used?

-lp limit the number of thread(logical processor).

**Pull request merged**